### ISSUE TITLE
json.hpp compilation issue with other typedefs with same name

### ISSUE THREAD

**Issue by kevinlul** [nlohmann/json#1642]
The issue is similar to #606 and #1530: a compilation error involving `json.hpp`, with `<` being interpreted as a template instead of "less than"

![nlohmann-json](https://user-images.githubusercontent.com/6320810/59569770-2a257680-905c-11e9-8671-946cd7b6e6ad.PNG)

Essentially, other typedefs with the same name as the ones used by the library are taken as reference when templating, which happens often under `using namespace`

This is confirmed to happen with GCC 7.4.0 on Ubuntu and GCC 9.1.0 on macOS Mojave, compiling against the version from `apt-get` and `brew` respectively.

A fix would be wrapping the comparisons with parentheses.

Also discovered by @edo9300 and @DyXEL

**Comment by nickaein** [comment]
Thanks for reporting the issue! 

This is a reoccurring issue (e.g. https://github.com/nlohmann/json/issues/590, https://github.com/nlohmann/json/issues/606, https://github.com/nlohmann/json/issues/1530) which  strangely, current CI builds haven't detected. For instance, the library are being built with GCC 4 through GCC 9 in [Travis CI](https://github.com/nlohmann/json/blob/develop/.travis.yml) yet this error remains uncaught.

Can you give more details on the build environment so the error can be reproduced and prevented in the future by setting up a regression build? It would be kind of you if you could reproduce the error on https://godbolt.org (you can set compiler version and build flags there).

**Comment by edo9300** [comment]
@nickaein i made a sample scenario where this happens https://godbolt.org/z/iHJVjA
```cpp
template <typename T>
class string {
};
#include <nlohmann/json.hpp>
int main() {
	return 0;
}
```
you can replace that templated class with any of the classes used by this library, and get the same error. Weirdly enough this doesn't happen when compiling for windows https://godbolt.org/z/aoAxd0

**Comment by nickaein** [comment]
@edo9300: Thanks for the example! We might have to add something like this as a regression test.

@kevinlul: I've just noticed you have already opened an PR for this issue. Thanks for that! You might want to add a **Fixes #1642** phrase ([or other keywords mentioned here](https://help.github.com/en/articles/closing-issues-using-keywords)) in the description of PR so that gets linked to this issue and automatically close the issue when merged.
